pkg/ingester/owned_series_test.go

```diff
@@ -0,0 +1,413 @@
+	cnt, sh := c.db.ownedSeriesAndShards()
+	require.Equal(t, series, cnt)
+	require.Equal(t, shards, sh)
+}
+
+func (c *ownedSeriesTestContext) updateOwnedSeriesAndCheckResult(t *testing.T, ringChanged bool, expectedUpdatedTenants int, expectedReason string) {
+	c.buf.Reset()
+	require.Equal(t, expectedUpdatedTenants, c.ownedSeries.updateAllTenants(context.Background(), ringChanged), c.buf.String())
+	require.Contains(t, c.buf.String(), expectedReason)
+}
+
+func (c *ownedSeriesTestContext) registerTestedIngesterIntoRing(t *testing.T, instanceID, instanceAddr, instanceZone string) {
+	c.ingesterZone = instanceZone
+
+	// Insert our ingester into the ring. When lifecycler starts, it will find this entry, and keep the tokens.
+	updateRing(t, c.kvStore, func(desc *ring.Desc) {
+		var tokens []uint32
+		for i := 0; i < ownedServiceSeriesCount/2; i++ {
+			tokens = append(tokens, c.seriesTokens[i]+1)
+		}
+		// This instance be the second ingester in the shuffle shard (skip=1). "second" ingester will be first ingester in the shard.
+		tokens = append(tokens, userToken(ownedServiceTestUser, instanceZone, 1)+1)
+		slices.Sort(tokens)
+
+		desc.AddIngester(instanceID, instanceAddr, instanceZone, tokens, ring.ACTIVE, time.Now())
+	})
+}
+
+// Insert second ingester to the ring, with tokens that will make it second half of the series.
+// This ingester will also be first ingester in the user's shuffle shard (skip: 0).
+func (c *ownedSeriesTestContext) registerSecondIngesterOwningHalfOfTheTokens(t *testing.T) {
+	updateRing(t, c.kvStore, func(desc *ring.Desc) {
+		var tokens []uint32
+		for i := ownedServiceSeriesCount / 2; i < ownedServiceSeriesCount; i++ {
+			tokens = append(tokens, c.seriesTokens[i]+1)
+		}
+		tokens = append(tokens, userToken(ownedServiceTestUser, c.ingesterZone, 0)+1)
+		slices.Sort(tokens)
+
+		// Must be in the same zone, because we use RF=1, and require RF=num of zones.
+		desc.AddIngester("second-ingester", "localhost", c.ingesterZone, tokens, ring.ACTIVE, time.Now())
+	})
+}
+
+func (c *ownedSeriesTestContext) removeSecondIngester(t *testing.T) {
+	updateRing(t, c.kvStore, func(desc *ring.Desc) {
+		desc.RemoveIngester("second-ingester")
+	})
+}
+
+func TestOwnedSeriesService(t *testing.T) {
+	// Generate some series, and compute their hashes.
+	var seriesToWrite []series
+	var seriesTokens []uint32
+	for seriesIdx := 0; seriesIdx < ownedServiceSeriesCount; seriesIdx++ {
+		s := series{
+			lbls:      labels.FromStrings(labels.MetricName, "test", fmt.Sprintf("lbl_%05d", seriesIdx), "value"),
+			value:     float64(0),
+			timestamp: time.Now().UnixMilli(),
+		}
+		seriesToWrite = append(seriesToWrite, s)
+		seriesTokens = append(seriesTokens, mimirpb.ShardByAllLabels(ownedServiceTestUser, s.lbls))
+	}
+
+	// Verify that series tokens have some gaps between them.
+	slices.Sort(seriesTokens)
+	for i := 1; i < len(seriesTokens); i++ {
+		require.Greater(t, seriesTokens[i]-seriesTokens[i-1], uint32(1))
```

Comment:
[nit] Isn't easier this 👇  way? Or am I missing something obvious?

```suggestion
		require.Greater(t, seriesTokens[i], seriesTokens[i-1])
```

Reply:
You're right. I was checking for larger gap previously.